### PR TITLE
Multigrid: Add short cycle

### PR DESCRIPTION
**Pull request by SuX97**
Add short cycle sampler.

### CODE REVIEW

File: mmaction/datasets/samplers/short_cycle_sampler.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from torch.utils.data.sampler import Sampler
+
+
+class ShortCycleBatchSampler(Sampler):
+    """Extend Sampler to support "short cycle" sampling. The Sampler input can
+    be both distributed or non-distributed.
+
+    See paper "A Multigrid Method for Efficiently Training Video Models", Wu et
+    al., 2019 (https://arxiv.org/abs/1912.00998) for details.
+    """
```
Comment (dreamerlin) [inline mmaction/datasets/samplers/short_cycle_sampler.py:12]: Args description is needed.

File: mmaction/datasets/builder.py [hunk 2]
```diff
@@ -71,6 +73,10 @@ def build_dataloader(dataset,
             Default: False
         pin_memory (bool): Whether to use pin_memory in DataLoader.
             Default: True
+        short_cycle (bool): Whether to use a short_cycle sampler.
+            Default: False.
+        multi_grid_cfg (dict): The config for multi-grid training.
```
Comment (dreamerlin) [inline mmaction/datasets/builder.py:78]: (dict | None)

File: mmaction/datasets/samplers/short_cycle_sampler.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        self.drop_last = drop_last
+
+        bs_factor = [
+            int(round(1 / s**2)) for s in multi_grid_cfg.short_cycle_factors
+        ]
+
+        self.batch_sizes = [
+            batch_size * bs_factor[0],
+            batch_size * bs_factor[1],
+            batch_size,
```
Comment (dreamerlin) [inline mmaction/datasets/samplers/short_cycle_sampler.py:42]: remove the last comma

File: mmaction/datasets/samplers/short_cycle_sampler.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            raise ValueError(
+                'sampler should be an instance of '
+                'torch.utils.data.Sampler, but got sampler={}'.format(sampler))
+        if (not isinstance(batch_size, _int_classes)
+                or isinstance(batch_size, bool) or batch_size <= 0):
+            raise ValueError('batch_size should be a positive integer value, '
+                             'but got batch_size={}'.format(batch_size))
+        if not isinstance(drop_last, bool):
+            raise ValueError('drop_last should be a boolean value, but got '
+                             'drop_last={}'.format(drop_last))
```
Comment (dreamerlin) [inline mmaction/datasets/samplers/short_cycle_sampler.py:31]: f-string

File: mmaction/datasets/samplers/short_cycle_sampler.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    def __init__(self, sampler, batch_size, drop_last, multi_grid_cfg):
+        if not isinstance(sampler, Sampler):
+            raise ValueError(
+                'sampler should be an instance of '
+                'torch.utils.data.Sampler, but got sampler={}'.format(sampler))
+        if (not isinstance(batch_size, _int_classes)
+                or isinstance(batch_size, bool) or batch_size <= 0):
+            raise ValueError('batch_size should be a positive integer value, '
+                             'but got batch_size={}'.format(batch_size))
```
Comment (dreamerlin) [inline mmaction/datasets/samplers/short_cycle_sampler.py:28]: f-string

File: mmaction/datasets/samplers/short_cycle_sampler.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        batch_size (int): The batchsize before short-cycle modification.
+        drop_last (bool): Whether to drop the last incomplete batch in epoch.
+        multi_grid_cfg (dict): The config dict for multi-grid trainingl.
+    """
+
+    def __init__(self, sampler, batch_size, drop_last, multi_grid_cfg):
+        if not isinstance(sampler, Sampler):
+            raise ValueError(
+                'sampler should be an instance of '
+                'torch.utils.data.Sampler, but got sampler={}'.format(sampler))
```
Comment (dreamerlin) [inline mmaction/datasets/samplers/short_cycle_sampler.py:24]: 1. f-string.
2. 'torch.utils.data.Sampler'

File: mmaction/datasets/rawframe_dataset.py [hunk 7]
```diff
@@ -86,11 +89,14 @@ def __init__(self,
...
                  start_index=1,
-                 modality='RGB'):
+                 modality='RGB',
+                 **kwargs):
         self.filename_tmpl = filename_tmpl
         self.with_offset = with_offset
         super().__init__(ann_file, pipeline, data_prefix, test_mode,
                          multi_class, num_classes, start_index, modality)
+        if kwargs.get('short_cycle_factors', None) is not None:
+            self.short_cycle_factors = kwargs['short_cycle_factors']
```
Comment (dreamerlin) [inline mmaction/datasets/rawframe_dataset.py:99]:
```
self.short_cycle_factors = kwargs.get('short_cycle_factors', None)
```

**Comment by dreamerlin** [comment]
[Unittest](https://codecov.io/gh/open-mmlab/mmaction2/pull/134/diff?src=pr&el=tree#diff-bW1hY3Rpb24vZGF0YXNldHMvcGlwZWxpbmVzL2Zvcm1hdGluZy5weQ==) is needed

**Comment by SuX97** [comment]
> [Unittest](https://codecov.io/gh/open-mmlab/mmaction2/pull/134/diff?src=pr&el=tree#diff-bW1hY3Rpb24vZGF0YXNldHMvcGlwZWxpbmVzL2Zvcm1hdGluZy5weQ==) is needed

Unittest added[here](https://github.com/open-mmlab/mmaction2/pull/29).

**Pull request closed without merging**